File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.py
# P의 좌표를 모으고, 각 좌표별로 BFS를 돌린다.
# P의 거리가 2 이하에 암것도 없으면 1
# 뭔가가 걸린다면 0
import copy
from collections import deque

def bfs(room, p_list):
    dx = [-1, 1, 0, 0]
    dy = [0, 0, 1, -1]
    for y, x in p_list:
        c_room = copy.deepcopy(room)
        dq = deque()
        dq.append((y,x,0))    
        c_room[y][x] = 'X'
        while dq:
            y, x, cnt = dq.popleft()
            if cnt >= 2:
                break
            for i in range(4):
                after_y = y + dy[i]
                after_x = x + dx[i]
                if after_y < 0 or after_y >= len(c_room) or after_x < 0 or after_x >= len(c_room[0]) or c_room[after_y][after_x] == 'X':
                    continue
                elif c_room[after_y][after_x] == 'P':
                    return 0
                else:
                    c_room[after_y][after_x] = 'X'
                    dq.append((after_y, after_x, cnt + 1))
    return 1
        

def solution(places):
    # room을 만든다 
    rooms = []
    res = []
    for place in places:
        p_list = []
        for p in place:
            p_list.append(list(p))
        rooms.append(p_list)
    
    for room in rooms:
        # p의 좌표 모으기
        p_list = []
        for y in range(len(room)):
            for x in range(len(room[0])):
                if room[y][x] == 'P':
                    p_list.append((y,x))
        val = bfs(room, p_list)
        res.append(val)
    
    return res

File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.java
import java.util.*;

class Solution {
    public int bfs(char[][] room, List<int[]> pList) {
        int[] dx = {-1, 1, 0, 0};
        int[] dy = {0, 0, 1, -1};
        
        for(int[] p : pList) {
            int startY = p[0];
            int startX = p[1];
            
            char[][] cRoom = new char[room.length][];
            
            for (int i = 0; i < room.length; i++) {
                cRoom[i] = room[i].clone();
            }

            Queue<int[]> q = new ArrayDeque<>();
            
            q.offer(new int[]{startY, startX, 0});
            
            cRoom[startY][startX] = 'X';
            
            while (!q.isEmpty()) {

                int[] cur = q.poll();

                int y = cur[0];
                int x = cur[1];
                int cnt = cur[2];

                // 맨해튼 거리 2까지만 확인
                if (cnt >= 2) {
                    break;
                }

                for (int i = 0; i < 4; i++) {

                    int afterY = y + dy[i];
                    int afterX = x + dx[i];

                    // 범위 밖이거나
                    // 파티션 또는 이미 방문한 곳이라면
                    if (
                        afterY < 0 || afterY >= cRoom.length ||
                        afterX < 0 || afterX >= cRoom[0].length ||
                        cRoom[afterY][afterX] == 'X'
                    ) {
                        continue;
                    }

                    // 다른 사람이 발견됨
                    if (cRoom[afterY][afterX] == 'P') {
                        return 0;
                    }

                    // 방문 처리
                    cRoom[afterY][afterX] = 'X';

                    q.offer(new int[]{
                        afterY,
                        afterX,
                        cnt + 1
                    });
                }
            }
        }

        return 1;
    }
    
    public int[] solution(String[][] places) {
        int[] res = new int[5];
        
        for (int i = 0; i < 5; i++) {
            String[] place = places[i];
            
            char[][] room = new char[5][];
            
            for (int j = 0; j < 5; j++) {
                room[j] = place[j].toCharArray();
            }
            
            List<int[]> pList = new ArrayList<>();
            
            for (int y = 0; y < 5; y++) {
                for (int x = 0; x < 5; x++) {
                    if (room[y][x] == 'P') {
                        pList.add(new int[]{y,x});
                    }
                }
            }
            res[i] = bfs(room, pList);
        }
        return res;
    }
}